### PR TITLE
CSS load from resource and add gresource_* functions

### PR DESCRIPTION
**Pull request by x1unix**
Added basic GResource methods:

**GTK**
* `gtk_css_provider_load_from_resource()`

**GIO**
- `g_resource_load()`
- `g_resource_new_from_data()`
- `g_resources_register()`
- `g_resources_unregister()`
- `g_resources_enumerate_children()`

Also added Go module for this library

### CODE REVIEW

**Comment by x1unix** [comment]
Currently I have no idea what caused these Travis build errors:

- [`gio/gresources/gresources.go:40:8: could not determine kind of name for C.free`](https://travis-ci.org/github/gotk3/gotk3/jobs/678497751)
- [`gtk/gtk.go:2907:2: could not determine kind of name for C.gtk_css_provider_load_from_resource`](https://travis-ci.org/github/gotk3/gotk3/jobs/678497750)

I've tested mentioned GResource functions on my GTK app on macOS, Windows 10 and Ubuntu 19 and everything works well (GTK 3)

**Comment by MJacred** [comment]
Hi, first of all: thanks for your contribution!

Here are some pointers:
* please check in which GTK version gtk_css_provider_load_from_resource() was introduced. you probably put it into the wrong file -> mind the build tags at the top
* you probably got the C import wrong -> you need to have linebreaks in a certain way -> look at other files
* please don't add a *.mod file -> this drops support for anyone with Golang version < 1.11

**Comment by x1unix** [comment]
@MJacred 

1. Unfortunately I can't find info about in which version that method appears (checked the [docs](https://developer.gnome.org/gtk3/stable/GtkCssProvider.html#GtkCssProvider.description) ).
2. I don't think that `go.mod` file should cause issues on some old Go versions (old `go get` will just ignore that file)

File: gio/gresources/gresource.go.h [hunk 1]
```diff
@@ -0,0 +1,8 @@
+#pragma once
+
+#include <stdint.h>
+#include <stdlib.h>
+#include <string.h>
+#include <stdio.h>
+
+static inline gchar** next_gcharptr(gchar** s) { return (s+1); }
```
Comment (MJacred) [inline gio/gresources/gresource.go.h:8]: please add a newline

File: .gitignore [hunk 2]
```diff
@@ -1 +1,2 @@
-./idea/*
\ No newline at end of file
+.idea
+.vscode
```
Comment (MJacred) [inline .gitignore:2]: please add a newline

File: gio/gresources/gresources.go [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+// #include <gio/gio.h>
+// #include <stdlib.h>
+import "C"
+import (
+	"errors"
+	"unsafe"
+)
+
+// GResourceLookupFlags is a representation of GTK's GResourceLookupFlags
+type LookupFlags int
```
Comment (MJacred) [inline gio/gresources/gresources.go:12]: please name it `ResourceLookupFlags`

**Comment by MJacred** [comment]
on second thought, you are right (regarding 2.)

**Comment by x1unix** [comment]
@MJacred 

> please change package name to gio

I've beed inspired by **PyGTK** and `gtk3` package for ruby, thats why I've made dedicated sub-package for whole gresources logic (to avoid `GResource` prefix for methods).

This is how it's made in ruby's `gtk3` or **PyGTK**:

```ruby
#!/usr/bin/env ruby

require "gtk3"

resource = Gio::Resource.load("simple_window.gresource")
Gio::Resources.register(resource)
builder = Gtk::Builder.new(:resource => "/simple_window/simple_window.ui")
window = builder.get_object("window")
window.show_all
Gtk.main
```

File: gio/gresources/gresource.go.h (hunk 1 above)
Comment (x1unix) [reply gio/gresources/gresource.go.h]: Done - https://github.com/gotk3/gotk3/pull/565/commits/fd392564be1bbbdfdab5a3357f27cc90f190b966

File: .gitignore (hunk 2 above)
Comment (x1unix) [reply .gitignore]: Done - https://github.com/gotk3/gotk3/pull/565/commits/fd392564be1bbbdfdab5a3357f27cc90f190b966

File: gio/gresources/gresources.go (hunk 3 above)
Comment (x1unix) [reply gio/gresources/gresources.go]: Done - https://github.com/gotk3/gotk3/pull/565/commits/31fcf3e164a4e6869b2ea420c9c8328ce0709eea

**Comment by x1unix** [comment]
@MJacred moved `gtk.CSSProvider.LoadFromResource()` under `gtk_since_3_16.go` - https://github.com/gotk3/gotk3/pull/565/commits/ec08a3a1c82ef3f7afc443679f3b4134e6b6cf77

**Comment by x1unix** [comment]
@MJacred moved `gresource` to `gio` - https://github.com/gotk3/gotk3/pull/565/commits/01713b8beedb7a4ef3d2a042725b4baadf9826e1

looks like only `**gchar` to `**char``casting left

**Comment by MJacred** [comment]
> @MJacred
> 
> > please change package name to gio
> 
> I've beed inspired by **PyGTK** and `gtk3` package for ruby, thats why I've made dedicated sub-package for whole gresources logic (to avoid `GResource` prefix for methods).
> 
> This is how it's made in ruby's `gtk3` or **PyGTK**:
> 
> ```ruby
> #!/usr/bin/env ruby
> 
> require "gtk3"
> 
> resource = Gio::Resource.load("simple_window.gresource")
> Gio::Resources.register(resource)
> builder = Gtk::Builder.new(:resource => "/simple_window/simple_window.ui")
> window = builder.get_object("window")
> window.show_all
> Gtk.main
> ```

I understand your reasoning. It's just that the whole gotk3 repo coding convention is like that. Breaking consistency here would confuse users...

On that note, you should rename your `func EnumerateChildren` (now that the package name is different).

**Comment by MJacred** [comment]
> @MJacred moved `gresource` to `gio` - [01713b8](https://github.com/gotk3/gotk3/commit/01713b8beedb7a4ef3d2a042725b4baadf9826e1)
> 
> looks like only `**gchar` to `**char`casting issue left
> 
> AFAIL under the hood, `gchar` is basically typedef alias to `char`, so this C code should work:
> 
> ```c
> char * gCharToChar(gchar* chr) {
>   return (char*) chr;
> }
> ```

you can rewrite your `toGoStringArray` to take `**char`. Because that's what `g_resources_enumerate_children` returns. Then you don't need to cast again in `C.GoString((*C.char)(*c))`. Problem is that `C.g_strfreev(originalc)` and `C.next_gcharptr(c)` will complain...

**Comment by x1unix** [comment]
@MJacred according to docs, `g_resources_enumerate_children` returns `**char` but requires to be freed as `**gchar`

image

**Comment by MJacred** [comment]
> @MJacred according to docs, `g_resources_enumerate_children` returns `**char` but requires to be freed as `**gchar`.


most peculiar

**Comment by MJacred** [comment]
> It's strange that casting error doesn't reproduce on my setup (macOS, Windows and Ubuntu with Go 1.13).

Hm... sometimes when I glance over go's changelog, I see changes to cgo or how go handles automatic dereferencing, so not that surprising if it suddenly works/doesn't work anymore.  
Not all travis build pipelines failed:
* only for go 1.8 and 1.9: https://travis-ci.org/github/gotk3/gotk3/builds/679389015, https://travis-ci.org/github/gotk3/gotk3/builds/679386164 (here go 1.13 failed due to different error)

**Comment by MJacred** [comment]
after you are finished, please clean up the commit history. Best with `hard reset`, I guess (example [here](https://docs.godotengine.org/en/latest/community/contributing/pr_workflow.html#updating-your-branch))

**Comment by x1unix** [comment]
@MJacred just select "Squash and Merge". This will squash all my changes into single commit.

**Comment by x1unix** [comment]
@MJacred looks like all builds are passing, also I've compacted all commits into 4 normal commits to keep clean history.

image

I hope this is all for now :)

**P.S** during my app development, I've created a tiny helper library that allows to bind components from `gtk.Builder` by ID into your view controller using `gtk` tag:

image

This saves development time and simplifies life.

Mb, this library can be added to `gotk3` main repo as sub-package (like `gotk3utils`)?

**Comment by MJacred** [comment]
Hm... not sure on that. You might want to open a new issue with this proposal. Maybe add a github snippet code via github gist (https://gist.github.com/). That would help understanding what you mean. I'm not an official maintainer of gotk3, so I can't make this decision anyway.

**Comment by x1unix** [comment]
@MJacred ok, nevermind :)

Can you merge this PR now? Or this should be done by official maintainer?

**Comment by MJacred** [comment]
Haha, I can't merge it as I'm not an official maintainer/member. You'll need to wait until one comes around. That's usually once or twice a month.

**Comment by x1unix** [comment]
@andre-hub @yamnikov-oleg can you review this PR please?

**Comment by andre-hub** [comment]
> @MJacred looks like all builds are passing, also I've compacted all commits into 4 normal commits to keep clean history.
> 
> image
> 
> I hope this is all for now :)
> 
> **P.S** during my app development, I've created a tiny helper library that allows to bind widgets from `gtk.Builder` by ID into your view controller using `gtk` tag:
> 
> image
> 
> This saves my development time alot.
> 
> Mb, this library can be added to `gotk3` main repo as sub-package (like `gotk3utils`)?

hi @x1unix 
i have create a new repository for optional functions. you can see it https://github.com/gotk3/gotk3utils

**Comment by andre-hub** [comment]
hi @x1unix 
hi @MJacred 

my current problem with this PR is that there is already a separate repository for the namespace gio. if i would now add this PR, there would be two locations for gio and thus later a mess. 

Therefore I would want to keep the separation. So I have to create a PR in the other repository.

[gotk3 gio](https://github.com/gotk3/gio)

**Comment by MJacred** [comment]
@andre-hub 
Indeed, there is one... I did not notice, because **the same gio bindings are already in this repository** (or at least most of them, I did not compare 1-to-1).

Hm... In my opinion, there should be no problem to include gio here, as gio is already in this repo and gio in general is kind of "part of" glib: The GTK team put glib + gio in one repo and gtk + gdk in another. Furthermore, glib and gio use the same version releases.  
If we do a split, it should be the same as GTK. But we can also do no splitting and get rid of the other repo. **What's your opinion?**

**Comment by x1unix** [comment]
@MJacred @andre-hub gotk3/gio repo wasn't update since 2016, so I thought it was abandoned and created functionality here (some gio related things already exists here).
Also whats the point of separated gio package? (rest of package are in main gotk3 package)

**Comment by MJacred** [comment]
Regarding GTK, my reasoning is: gio and glib can be used without gtk and gdk, because the former is purely logic, while the latter is GUI stuff.  
In gotk3, I assume because gio is often not used in most (GUI) applications...

**Comment by x1unix** [comment]
@MJacred excuse me, but I can't imagine who will need to use gio without GTK.
gio allows to use GResources (I can bundle all layout files, styles and images in single gresource instead of taking all that files with executable)

**Comment by MJacred** [comment]
you should read https://gitlab.gnome.org/GNOME/glib/

GTK+ team bundled it like this

first repo
* GTK
* ATK
* GDK
* GSK

second repo
* GLib
* Gio

not sure where cairo and pixbuf are

**Comment by x1unix** [comment]
@MJacred unfortunately, `gio` package is broken now

изображение

**Pull request merged**